pkg/shell/state.jsx

```diff
@@ -0,0 +1,583 @@
+     */
+
+    let language = document.cookie.replace(/(?:(?:^|.*;\s*)CockpitLang\s*=\s*([^;]*).*$)|^.*$/, "$1");
+    if (!language)
+        language = navigator.language.toLowerCase(); // Default to Accept-Language header
+
+    const config = {
+        language,
+        language_direction: cockpit.language_direction,
+        host_switcher_enabled: false,
+    };
+
+    /* Host switcher enabled? */
+    const meta_multihost = document.head.querySelector("meta[name='allow-multihost']");
+    if (meta_multihost instanceof HTMLMetaElement && meta_multihost.content == "yes")
+        config.host_switcher_enabled = true;
+
+    /* MACHINES DATABASE AND MANIFEST LOADER
+     *
+     * These are part of the machinery in the basement that maintains
+     * the database of all hosts (including "localhost", and monitors
+     * their manifests.
+     */
+
+    const machines = machines_factory.instance();
+    const loader = machines_factory.loader(machines);
+
+    machines.addEventListener("ready", on_ready);
+
+    machines.addEventListener("removed", (ev, machine) => {
+        remove_machine_frames(machine);
+    });
+    machines.addEventListener("added", (ev, machine) => {
+        preload_machine_frames(machine);
+    });
+    machines.addEventListener("updated", (ev, machine) => {
+        if (!machine.visible || machine.problem)
+            remove_machine_frames(machine);
+        else
+            preload_machine_frames(machine);
+    });
+
+    if (machines.ready)
+        on_ready();
+
+    function on_ready() {
+        if (machines.ready) {
+            self.ready = true;
+            window.addEventListener("popstate", ev => {
+                update();
+                ensure_frame_loaded();
+                ensure_connection();
+            });
+
+            update();
+            ensure_frame_loaded();
+            ensure_connection();
+        }
+    }
+
+    /* WATCH DOGS
+     */
+
+    const watchdog = cockpit.channel({ payload: "null" });
+    watchdog.addEventListener("close", (event, options) => {
+        const watchdog_problem = options.problem || "disconnected";
+        console.warn("transport closed: " + watchdog_problem);
+        self.problem = watchdog_problem;
+        // We might get here real early, before events seem to
+        // work. Let's push the update processing to the event loop.
+        setTimeout(() => update(), 0);
+    });
+
+    const old_onerror = window.onerror;
+    window.onerror = function cockpit_error_handler(msg, url, line) {
+        // Errors with url == "" are not logged apparently, so let's
+        // not show the "Oops" for them either.
+        if (url != "") {
+            self.has_oops = true;
+            update();
+        }
+        if (old_onerror)
+            return old_onerror(msg, url, line);
+        return false;
+    };
+
+    /* FRAMES
+     *
+     * Frames are created on-demand when navigating to them for the
+     * first time, by calling ENSURE_FRAME.
```

Comment:
The name is `ensure_frame()`. Would be good to be correct, for `grep`.

Reply:
Yes, changed. Also in other places.